eksrollup/lib/k8s.py

```diff
@@ -152,6 +152,144 @@ def taint_node(node_name):
 
 
+def deployment_restart(node_name):
+    """
+    Availability is not guaranteed for a single replica deployment.
+    PDB only works with more than 1 replica. Node draining without
+    PDB is synchronous, meaning new pods are started at the same
+    time old pods are being terminated. A deployment restart will
+    ensure there is always one pod available to the service. If
+    enabled, this function will only perform a rolling restart on
+    deployments with single replicas.
+
+    This can raise an Kubernetes API exception, we're happy to let it bubble up
+    to the main caller and for the logic to terminate.
+    """
+
+    # Get the applicable deployments, this will return a list of tuples or raise an 
+    # exception that we can let bubble up to the main caller
+    deployments = find_applicable_deployments(node_name)
+    if len(deployments) <= 0:
+        logger.info("No single-replica deployment found for node %s", node_name)
+        return
+
+    # Will track the sum of all return codes
+    total_return_codes = 0
+
+    with ThreadPoolExecutor(max_workers=10) as executor:
+        futures = []
+        for (deployment_name, deployment_namespace) in deployments:
+            future = executor.submit(rolling_restart, deployment_name, deployment_namespace)
+            futures.append(future)
+
+        # Attach threads
+        for thr in as_completed(futures):
+            returncode = thr.result()
+            total_return_codes += returncode
+
+        logger.info("Deployment restart function returned %s errors", total_return_codes)
+        if total_return_codes != 0:
+            # Raise our own exception to let the caller know we failed to restart all applicable pods
+            raise Exception("Deployment restart failed with %s errors", total_return_codes)
+
+
+def find_applicable_deployments(node_name):
+    """
+    Find the kubernetes deployment (replicaset) that have a single pod running.
+
+    This can raise an Kubernetes API exception, we're happy to let it bubble up
+    to the main caller and for the logic to terminate.
+    """
+    deployments = []
+
+    ensure_config_loaded()
+
+    k8s_api = client.CoreV1Api()
+    apps_k8s_api = client.AppsV1Api()
+
+    # Node field selector used to discover all pods on the node
+    node_field_selector = f"spec.nodeName={node_name}"
+
+    # List all pods on the node
+    pods = k8s_api.list_pod_for_all_namespaces(field_selector=node_field_selector)    
+    for pod in pods.items:
+        pod_name = pod.metadata.name
+        namespace = pod.metadata.namespace  # namespace of the pod
+        # Using ownerReferences to verify the pod is owned by a ReplicaSet
+        for pod_owner_reference in pod.metadata.owner_references:
+            if pod_owner_reference.kind == "ReplicaSet":
+                rs_name = pod_owner_reference.name
+
+                # Get the ReplicaSet of the pod
+                rs = apps_k8s_api.read_namespaced_replica_set(rs_name, namespace)
```

Comment:
Now, if an exception happens during the call to `apps_k8s_api`, would it be caught and appended to the futures thread, while the loop continues?

Reply:
Well this happens before / outside of the threads from the main process - it could raise an exception but the calling function is prepared to handle it - see cli.py:219
```
        for outdated in outdated_instances:
            # catch any failures so we can resume aws autoscaling
            try:
                # get the k8s node name instead of instance id
                node_name = get_node_by_instance_id(k8s_nodes, outdated['InstanceId'])
                desired_asg_capacity -= 1
                if app_config['RESTART_DEPLOYMENT']:
                    deployment_restart(node_name)
                drain_node(node_name)
                delete_node(node_name)
                # [...]
            except Exception as drain_exception:
                logger.info(drain_exception)
                raise RollingUpdateException("Rolling update on ASG failed", asg_name)
```

Reply:
I tested to see what happens if `rs = apps_k8s_api.read_namespaced_replica_set(rs_name, namespace)` returned an error, and it doesn't seem the exception is caught. All I did was  change the namespace purposely only for 1 pod:

```
        if pod_name == "crypto-mgmt-service-549795f48c-4t8w7":
            namespace = "test"
```

Reply:
I would say that it’s just return no result in that case and not raise an exception for that…

Reply:
Yeah it did exit with "kubernetes.client.rest.ApiException: (404)". Did we want to exit the script when the api returns a 404? Or should we log the error and continue the loop? Unless we are purposely existing on exception?

Reply:
Awesome, tl'dr we want to exit on api error since we do not want to proceed on exception.